### PR TITLE
OLD - T350782 Test setup done in JS / WDIO vs Shell scripts

### PR DESCRIPTION
**Pull request by lorenjohnson**
Makes possible running tests more atomically, and more. https://phabricator.wikimedia.org/T350782

* Replaces all test runner shell scripts and with Typescript making more clear and flexible control of the our test environment
* Update test suite setup to allow for a full set of overrides to the test setup, and running process. See `*.conf.js` files and review the related Typescript types which self-document much of what is happening. 
* Provides detailed WDIO test run logs in `suites/SUITE_NAME/results/wdio*.log`. Here we can see all browser console output, as well as detail results of each selector happening in each spec file. Setup log from steps before test runners start is `wdio.log`, subsequent log files are created for each spec file ran according to the correlated WDIO job runner number, e.g. `wdio-0-1.log`, etc. 
* Changes local dev test run CLI from `make test SUITE=<suite-name>` to `./test.sh <suite-name>`
* Adds `--setup` test runner option to bring-up a test suites services without running tests, and leave them up along with the Selenium instance for manual testing, e.g `./test.sh quickstatements --setup`
* Allows all WDIO Runner options to be added to test runs :`./test.sh <suite-name> <WDIO-Runner-CLI-options>`. E.g. `./test.sh repo --spec=specs/repo/property.ts --watch`, etc
* Rewrites Example test setup to be in line with other tests. The Example test suite is now ran using simply: `./test.sh example`
* Rewrites Upgrade test setup to be in line with other tests. Run the Upgrade test suite with: `./test.sh upgrade WMDE9 <optional-version-tag-of-upgrade-to-version>`
* Renames all `*_SERVER` env variables to `*_URL` in the test environment (the only place we were using this variables) and adds the `http://` protocol prefix where it wasn't consistently applied before.
* Removes linting run before test runs
* Fix: Removes password requirement on Selenium Grid / Test Browser instance (docs updated)
* Fix: Linting no longer tries to lint Dockerfiles within `node_modules` directories
* Fix: In Github Actions removes `lint` as a "need" of the `build` and `test` jobs . This will have the effect of the checks still failing if linting fails, but `build` and `test` continue running. 
... etc

---

Please try out the new Test CLI, here is the help for that which can be retrieved by simply typing `./test.sh` or `./test.sh --help`:

```
> ./test.sh --help

./test.sh <suiteName>

Run a test suite

Commands:
  ./test.sh <suiteName>                        Run a test suite                                                [default]
  ./test.sh upgrade <fromVersion> [toVersion]  run upgrade test
  ./test.sh all                                Run all test suites

Positionals:
  suiteName  Name of the test suite to run or setup
   [string] [choices: "repo", "fedprops", "repo_client", "quickstatements", "pingback", "confirm_edit", "elasticsearch",
                                       "base__repo", "base__repo_client", "base__pingback", "base__fedprops", "example"]

Options:
  --setup                  Start and leave-up the test environment without running tests                       [boolean]
  --headedTests, --headed  Run tests in a headed browser                                                       [boolean]
  --spec                   Run only this spec file(s) using the selected test suite setup (WDIO)                 [array]
  --watch                  Automatically re-run tests when spec files in the suite change (WDIO)               [boolean]
  --version                Show version number                                                                 [boolean]
  --help                   Show help                                                                           [boolean]
```

Also help is available specific to the Upgrade test suite:

```
> ./test.sh upgrade --help

./test.sh upgrade <fromVersion> [toVersion]

run upgrade test

Positionals:
  fromVersion  Version to upgrade FROM in the upgrade test
  [string] [required] [choices: "WMDE9", "WMDE9_BUNDLE", "WMDE10", "WMDE10_BUNDLE", "WMDE11", "WMDE11_BUNDLE", "WMDE12",
                                        "WMDE12_BUNDLE", "WMDE13", "WMDE13_BUNDLE", "LOCAL_BUILD", "LOCAL_BUILD_BUNDLE"]
  toVersion    Version to upgrade TO in the upgrade test
  [string] [choices: "WMDE9", "WMDE9_BUNDLE", "WMDE10", "WMDE10_BUNDLE", "WMDE11", "WMDE11_BUNDLE", "WMDE12", "WMDE12_BU
                                                  NDLE", "WMDE13", "WMDE13_BUNDLE", "LOCAL_BUILD", "LOCAL_BUILD_BUNDLE"]

Options:
  --setup                  Start and leave-up the test environment without running tests                       [boolean]
  --headedTests, --headed  Run tests in a headed browser                                                       [boolean]
  --spec                   Run only this spec file(s) using the selected test suite setup (WDIO)                 [array]
  --watch                  Automatically re-run tests when spec files in the suite change (WDIO)               [boolean]
  --version                Show version number                                                                 [boolean]
  --help                   Show help                                                                           [boolean]
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -34,9 +34,9 @@ $ ./build.sh --no-cache wdqs
 ### Test Commands
 
 ```
-$ make test
-$ make test SUITE=repo
-$ make test SUITE=repo FILTER=special-item
+$ ./test.sh
+$ ./test.sh repo
+$ ./test.sh repo --spec special-item
 $ make test-upgrade VERSION=wmde.9
```
Comment (lorenjohnson) [inline README.md:40]: Update this.

File: docs/topics/testing.md [hunk 2]
```diff
@@ -31,29 +31,29 @@ Tests upgrading between a previous release defined in `test/suites/upgrade/old-v
 make test-upgrade VERSION=wmde.0
 ```
```
Comment (lorenjohnson) [inline docs/topics/testing.md:32]: This needs to change.

File: docs/topics/pipeline.md [hunk 3]
```diff
@@ -102,11 +98,15 @@ GZIP_COMPRESSION_RATE=1
...
+#### Timeout for mocha tests (in milliseconds)
+```
+MOCHA_OPTS_TIMEOUT=90000
 ```
-#### Timeout for mocha tests
+#### Timeout for waitFor* calls in tests (in milliseconds)
 ```
-MOCHA_OPTS_TIMEOUT=90
+WAIT_FOR_TIMEOUT=30000
 ```
```
Comment (lorenjohnson) [inline docs/topics/pipeline.md:110]: Check docs on timeouts.

File: test/helpers/default-functions.ts [hunk 4]
```diff
@@ -70,52 +79,28 @@ export function defaultFunctions(): void {
...
 	/**
-	 * Get installed extensions on wiki
+	 * Skip test if extension is not installed (present) on the Wikibase server
 	 */
 	browser.addCommand(
-		'getInstalledExtensions',
-		async ( server: string ): Promise<string[] | undefined> => {
-			const result = await browser.makeRequest(
-				`${server}/w/api.php?action=query&meta=siteinfo&siprop=extensions&format=json`
+		'skipIfExtensionNotPresent',
```
Comment (RickiJay-WMDE) [inline test/helpers/default-functions.ts:85]: Need to update `wdio.d.ts`

**Review by RickiJay-WMDE — CHANGES_REQUESTED** [review]

File: test/helpers/default-functions.ts (hunk 4 above)
Comment (lorenjohnson) [reply test/helpers/default-functions.ts:85]: It is up to date with this though. What am I missing?

Comment (RickiJay-WMDE) [reply test/helpers/default-functions.ts:85]: Updated, my bad

**Comment by lorenjohnson** [comment]
Re-wrote this changeset into 8 commits in https://github.com/wmde/wikibase-release-pipeline/pull/542. If you haven't already invested in reviewing here, please review there. It may help to walk through each commit.

**Pull request closed without merging**